### ISSUE TITLE
Window shadows?

### ISSUE THREAD

**Issue by last-partizan** [PolyMeilex/sctk-adwaita#4]
Hello.

Recently sctk-adwaita support was added to winit (https://github.com/rust-windowing/winit/issues/1967), and i checked it, looks great. Thank you for this project.

Is there a way to add shadows for a window?

**Comment by PolyMeilex** [comment]
Technically, it's possible, I simply did not have enough motivation to work on it, and opted for a simple 1px border instead.

We would have to draw a gradient on all 4 edges, cache it, so it does not get redrawn on every resize (we are software rendered), and update only edges that were affected by the resize.
Another good optimization would be to just draw 1 strip of a gradient and tile/repeat it on the whole width of a window.
As for edges, I'm not sure how to do it, most likely a radial gradient.

**Comment by last-partizan** [comment]
I'd take a look at this.

Is there any good docs on how to draw gradients? (i have some expirience with rust, just underlated to rendering things).

**Comment by PolyMeilex** [comment]
You would most likely use [tiny_skia::Shader](https://docs.rs/tiny-skia/latest/tiny_skia/enum.Shader.html#) with tiny skia [fill_rect()](https://docs.rs/tiny-skia/latest/tiny_skia/struct.Pixmap.html#method.fill_rect)

You'd have to render that in top surface/texture https://github.com/PolyMeilex/sctk-adwaita/blob/0cae30e7bf6ec88c0f42e285703acb44bdf2e907/src/lib.rs#L401-L410
Left one: https://github.com/PolyMeilex/sctk-adwaita/blob/0cae30e7bf6ec88c0f42e285703acb44bdf2e907/src/lib.rs#L521-L540
Right one: https://github.com/PolyMeilex/sctk-adwaita/blob/0cae30e7bf6ec88c0f42e285703acb44bdf2e907/src/lib.rs#L564-L575
Bottom one: https://github.com/PolyMeilex/sctk-adwaita/blob/0cae30e7bf6ec88c0f42e285703acb44bdf2e907/src/lib.rs#L470-L478

Basically, there is a texture per edge.

**Comment by last-partizan** [comment]
Currently i don't have much time to spare for this, but if anyone wants to give it a try, it would be great.

**Comment by lictex** [comment]
> It's cumbersome to make comfortable shadows, I thought, maybe it would be easier to pre-render and hardcode into the code? I think it can be implemented by referring to here: https://docs.cocos.com/creator/2.0/manual/en/ui/sliced-sprite.html

i tried this ([here](https://github.com/lictex/sctk-adwaita/tree/shadow)) and it seems to work for basic cases but
* for small windows there are some badly overlaps
* dragging area for resizing becomes *super large*
* 12 `draw_pixmap()` calls per draw. not sure how expensive this is, at least i dont feel its more laggy than before
* might be blurry on hidpi displays (shadows shouldnt be affected much by this though)

idk how to make these better really (especially the first two, which makes me think this is actually.. unusable)

| | |
|----|----|
|![image](https://github.com/PolyMeilex/sctk-adwaita/assets/13201009/4687e262-dc31-4667-8ce1-e2de3488bd8a)|![image](https://github.com/PolyMeilex/sctk-adwaita/assets/13201009/112ee866-9c71-47fa-bdf0-bf3479982671)|

**Comment by last-partizan** [comment]
It looks good, and if that issue with resize handles can be resolved, it would be usable.

I think issue with small windows can be ignored for now.

Looks like libdecor is ready: https://github.com/kovidgoyal/kitty/issues/3284#issuecomment-1536538321, but i haven't seen rust integration. So, we can use this until we have better options.

**Comment by Tim-Paik** [comment]
I seem to have a little idea, maybe we can set a threshold, and when it is less than this threshold, only the shadows of the four corners are drawn instead of the shadows of the four sides, so as to prevent the overlap problem, can this be achieved? 

Or we can be more straightforward and stop drawing shadows directly when the window is too small.

**Comment by Friz64** [comment]
I had a go at this in #43, feedback appreciated.
